clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,132 @@
+    perf,
+    "unnecessary call to `HashSet::contains` followed by `HashSet::insert`"
+}
+declare_lint_pass!(HashsetInsertAfterContains => [HASHSET_INSERT_AFTER_CONTAINS]);
+
+impl<'tcx> LateLintPass<'tcx> for HashsetInsertAfterContains {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if !expr.span.from_expansion()
+        && let Some(higher::If {
+            cond: cond_expr,
+            then: then_expr,
+            ..
+        }) = higher::If::hir(expr)
+        && let Some(contains_expr) = try_parse_contains(cx, cond_expr)
+        && find_insert_calls(cx, &contains_expr, then_expr) {
+            span_lint_and_then(
+                cx,
+                HASHSET_INSERT_AFTER_CONTAINS,
+                expr.span,
+                "usage of `HashSet::insert` after `HashSet::contains`",
+                |diag| {
+                    diag.note("`HashSet::insert` returns whether it was inserted")
+                        .span_help(contains_expr.span, "remove the `HashSet::contains` call");
+                },
+            );
+        }
+    }
+}
+
+struct ContainsExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+    span: Span,
+}
+fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Option<ContainsExpr<'tcx>> {
+    let expr = peel_hir_expr_while(expr, |e| match e.kind {
+        ExprKind::Unary(UnOp::Not, e) => Some(e),
+        _ => None,
+    });
+    if let ExprKind::MethodCall(
+        path,
+        receiver,
+        [
+            Expr {
+                kind: ExprKind::AddrOf(_, _, value),
+                span: value_span,
+                ..
+            },
+        ],
+        span,
+    ) = expr.kind
+    {
+        let receiver_ty = cx.typeck_results().expr_ty(receiver);
+        if value_span.ctxt() == expr.span.ctxt()
+            && is_type_diagnostic_item(cx, receiver_ty, sym::HashSet)
+            && path.ident.name == sym!(contains)
+        {
+            return Some(ContainsExpr { receiver, value, span });
+        }
+    }
+    None
+}
+
+struct InsertExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+}
+fn try_parse_insert<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<InsertExpr<'tcx>> {
+    if let ExprKind::MethodCall(path, receiver, [value], _) = expr.kind {
+        let receiver_ty = cx.typeck_results().expr_ty(receiver);
+        if is_type_diagnostic_item(cx, receiver_ty, sym::HashSet) && path.ident.name == sym!(insert) {
+            Some(InsertExpr { receiver, value })
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+fn find_insert_calls<'tcx>(cx: &LateContext<'tcx>, contains_expr: &ContainsExpr<'tcx>, expr: &'tcx Expr<'_>) -> bool {
+    for_each_expr(expr, |e| {
+        if let Some(insert_expr) = try_parse_insert(cx, e)
+        && SpanlessEq::new(cx).eq_expr(contains_expr.receiver, insert_expr.receiver)
```

Comment:
You should probably call `peel_borrows` before here so that `&x.contains(y)` and `x.insert(y)` are considered the same.